Cube arithmetic array type handling

rcomer opened the pull request:
Edit 20/07/22:

Closes #2987. During the discussion below we also established there are problems for the in-place case when the rhs object is more complex (i.e. masked or lazy) than the lhs cube’s data array.  So the current change also addresses that.

~This relates to #2987 but that issue is now fixed in numpy (see https://github.com/SciTools/iris/pull/3790#issuecomment-1184495096).  During the discussion below we also established there are problems for the in-place case when the rhs object is more complex (i.e. masked or lazy) than the lhs cube’s data array.  So the current change addresses that.  Also adds a test for the #2987 case, marked xfail until numpy 1.24 is released.~

~Although I'm targetting cube arithmetic here, this change will also have some effect on `apply_ufunc` and `IFunc`, for which it probably doesn't make sense: `apply_ufunc` would, I think, always expect numpy arrays, `IFunc` may be the same given that the function is user-determined.  I *think* the worst that will happen here is an inefficiency caused by converting a numpy array to a dask array and back again before the operation happens.  Does this inefficiency matter?  If so, the new behaviour could become an option in `_binary_op_common` that is explicitly chosen by the arithmetic functions.  For now, I've just gone for the simple approach.~

~I've yet to add a whatsnew as I'm not sure where it should go at the moment.   Might become more obvious once the Iris 3 release dust has settled...~

rcomer commented on lib/iris/tests/test_basic_maths.py:
```diff
@@ -685,12 +685,12 @@ def setUp(self):
+        self.cube_1i = iris.cube.Cube(self.data_1i.copy())
+        self.cube_2i = iris.cube.Cube(self.data_2i.copy())
+        self.cube_1u = iris.cube.Cube(self.data_1u.copy())
+        self.cube_2u = iris.cube.Cube(self.data_2u.copy())
```
Added these copy's because I concluded that previously the in-place tests were not very good.  Under Iris2.4, we have:
```python
import iris
import numpy as np

data_1f = np.array([[9, 9, 9], [8, 8, 8]], dtype=np.float64)
data_2f = np.array([[3, 3, 3], [2, 2, 2]], dtype=np.float64)

cube_1f = iris.cube.Cube(data_1f)
cube_2f = iris.cube.Cube(data_2f)

cube_1f += cube_2f
data_1f += data_2f

print(data_1f)
print(cube_1f.data is data_1f)
```
gives:
```
[[15. 15. 15.]
 [12. 12. 12.]]
True
```

rcomer commented:
Since @bjlittle has added this to the v3 milestone :grin: , I've gone ahead and added a line in whatsnew/latest.rst.  I think this is in line with the [new guidelines](https://github.com/SciTools/iris/blob/master/docs/iris/src/developers_guide/documenting/whats_new_contributions.rst).

rcomer commented:
Thanks @bjlittle, I've retargetted the branch.  In answer to your questions:

1. **Currently**, we only lose the mask for `real` and `lazy`, in that order.  **Previously** we did lose it the other way around, but it looks like @pp-mo instigated a fix in `dask` for that: https://github.com/dask/dask/issues/4441.  I wrote the new tests before I realised that change had happened, so possibly I am testing more than I need to.

2.  I have to admit that I hadn't given much attention to the in-place version, but now I have! In-place results are as above.

3. Yes.

4. Yes.

5. **Currently** the result is lazy *only if the first input cube is lazy*.

A script is worth a thousand words:

```python
import itertools

import iris
import numpy.ma as ma
import dask
import dask.array as da

print('Iris version', iris.__version__)
print('Dask version', dask.__version__)


def multiply_example(first_lazy=False, second_lazy=False, in_place=False):

    cube1 = iris.cube.Cube(da.ma.masked_array(range(5), mask=[0, 1, 1, 0, 0]))
    cube2 = iris.cube.Cube(da.ma.masked_array(range(5), mask=[0, 0, 1, 0, 1]))
    
    if first_lazy:
        str1 = "First lazy, "
    else:
        str1 = "First real, "
        cube1.data
        
    if second_lazy:
        str2 = "second lazy."
    else:
        str2 = "second real."
        cube2.data
        
    msg = str1 + str2
    
    if in_place:
        cube1 *= cube2
        result = cube1
        msg += ' Checking in-place'
    else:
        result = cube1 * cube2
        
    print(msg)
    print('Lazy result:', result.has_lazy_data())
    print('Result:', result.data)
    print('')
   
   
for in_place, first_lazy, second_lazy in itertools.product([False, True], repeat=3):
    multiply_example(first_lazy, second_lazy, in_place)
```

With `cube-arithmetic-post` branch, this gives:
```
Iris version 3.0.dev0
Dask version 2.20.0
First real, second real.
Lazy result: False
Result: [0 -- -- 9 --]

First real, second lazy.
Lazy result: False
Result: [0 -- -- 9 16]

First lazy, second real.
Lazy result: True
Result: [0 -- -- 9 --]

First lazy, second lazy.
Lazy result: True
Result: [0 -- -- 9 --]

First real, second real. Checking in-place
Lazy result: False
Result: [0 -- -- 9 --]

First real, second lazy. Checking in-place
Lazy result: False
Result: [0 -- -- 9 16]

First lazy, second real. Checking in-place
Lazy result: True
Result: [0 -- -- 9 --]

First lazy, second lazy. Checking in-place
Lazy result: True
Result: [0 -- -- 9 --]
```
With my branch it gives:
```
Iris version 3.0.dev0
Dask version 2.20.0
First real, second real.
Lazy result: False
Result: [0 -- -- 9 --]

First real, second lazy.
Lazy result: True
Result: [0 -- -- 9 --]

First lazy, second real.
Lazy result: True
Result: [0 -- -- 9 --]

First lazy, second lazy.
Lazy result: True
Result: [0 -- -- 9 --]

First real, second real. Checking in-place
Lazy result: False
Result: [0 -- -- 9 --]

First real, second lazy. Checking in-place
Lazy result: True
Result: [0 -- -- 9 --]

First lazy, second real. Checking in-place
Lazy result: True
Result: [0 -- -- 9 --]

First lazy, second lazy. Checking in-place
Lazy result: True
Result: [0 -- -- 9 --]

```
Probably I need to add tests for the in-place case.

rcomer commented:
> Probably I need to add tests for the in-place case.

Done

rcomer commented:
> 5. Currently the result is lazy only if the first input cube is lazy.

This statement appears only to be true if the first cube's data is type `MaskedArray`.  If you multiply a plain old numpy array by a dask array, you get a dask array.  So maybe this is a bug in `numpy.ma`?  In any case, my branch fixes it within Iris.

rcomer commented:
Also...

you can't do in-place arithmetic between numpy arrays and dask arrays.

```
In [3]: np.arange(3) += da.array([1, 2, 3])
  File "<ipython-input-3-328cd9b3455c>", line 1
    np.arange(3) += da.array([1, 2, 3])
                                       ^
SyntaxError: can't assign to function call
```
So I think some form of this branch would be needed regardless of the mask situation.

rcomer commented:
Reckon I'm gonna stop digging now! :scream: :flushed:

bjlittle commented:
@rcomer Awesome work 😀

Just curious... For the **current** branch, do you get that same behaviour but with regular NumPy arrays? (no masked arrays)

bjlittle commented:
@rcomer For me this issue seems to be in `dask` rather than`iris`, but we can work around it here for now... Ultimately we'd want to track this down and fix `dask` for a community win (if that is indeed where they issue lies)... But that's for another day 😉

rcomer commented:
OK, so I changed my script so the _first_ cube is not a masked type.  Also added exception handling because you just can't do in-place arithmetic between a numpy array and a dask array (which seems fair enough):

```python
import itertools

import iris
import dask
import dask.array as da

print('Iris version', iris.__version__)
print('Dask version', dask.__version__)


def multiply_example(first_lazy=False, second_lazy=False, in_place=False):

    cube1 = iris.cube.Cube(da.array(range(5)))
    cube2 = iris.cube.Cube(da.ma.masked_array(range(5), mask=[0, 0, 1, 0, 1]))
    
    if first_lazy:
        str1 = "First lazy, "
    else:
        str1 = "First real, "
        cube1.data
        
    if second_lazy:
        str2 = "second lazy."
    else:
        str2 = "second real."
        cube2.data
        
    msg = str1 + str2
    
    if in_place:
        try:
            cube1 *= cube2
        except NotImplementedError as e:
            print(e)
            print('')
            return
        
        result = cube1
        msg += ' Checking in-place'
    else:
        result = cube1 * cube2
        
    print(msg)
    print('Lazy result:', result.has_lazy_data())
    print('Result:', result.data)
    print('')
   
   
for in_place, first_lazy, second_lazy in itertools.product([False, True], repeat=3):
    multiply_example(first_lazy, second_lazy, in_place)
```
Under `cube-arithmetic-post`, this gives:
```
Iris version 3.0.dev0
Dask version 2.20.0
First real, second real.
Lazy result: False
Result: [0 1 -- 9 --]

First real, second lazy.
Lazy result: True
Result: [0 1 -- 9 --]

First lazy, second real.
Lazy result: True
Result: [0 1 -- 9 --]

First lazy, second lazy.
Lazy result: True
Result: [0 1 -- 9 --]

First real, second real. Checking in-place
Lazy result: False
Result: [ 0  1  4  9 16]

The out parameter is not fully supported. Received type ndarray, expected Dask Array

First lazy, second real. Checking in-place
Lazy result: True
Result: [0 1 -- 9 --]

First lazy, second lazy. Checking in-place
Lazy result: True
Result: [0 1 -- 9 --]
```

My branch lets us do the in-place real * lazy sensibly.  But note also that the `np.array * ma.array` in-place case (real * real) also strips the mask :scream:.  My branch doesn't do anything for that.

If we also change `cube2 = iris.cube.Cube(da.array(range(5)))`, we get this under `cube-arithmetic-post`:

```
Iris version 3.0.dev0
Dask version 2.20.0
First real, second real.
Lazy result: False
Result: [ 0  1  4  9 16]

First real, second lazy.
Lazy result: True
Result: [ 0  1  4  9 16]

First lazy, second real.
Lazy result: True
Result: [ 0  1  4  9 16]

First lazy, second lazy.
Lazy result: True
Result: [ 0  1  4  9 16]

First real, second real. Checking in-place
Lazy result: False
Result: [ 0  1  4  9 16]

The out parameter is not fully supported. Received type ndarray, expected Dask Array

First lazy, second real. Checking in-place
Lazy result: True
Result: [ 0  1  4  9 16]

First lazy, second lazy. Checking in-place
Lazy result: True
Result: [ 0  1  4  9 16]
```

rcomer commented:
So the problems are specific to:

* `real masked array * lazy masked array` - strips lazy mask.  Since `real plain array * lazy masked array` works fine, it seems probable that this can be fixed in `dask` or `np.ma`.  Note that this was discussed at https://github.com/dask/dask/issues/4441 and they thought there that the problem wasn't in `dask`.
* in-place arithmetic between a real plain array and a real masked array gives a real plain array.  I.e. the mask is lost. :flushed:

rcomer commented:
So I've just realised that my new in-place tests aren't testing the part of my code change that I thought they were, because we have two cubes and therefore `skeleton_cube` is `True`. :pensive:

rcomer commented:
Added some handling for the in-place real plain * masked case.  But I kinda feel like I'm making a mess... :cold_sweat:

rcomer commented:
Added some tests to cover the `not skeleton_cube` branch in `_math_op_common`.

rcomer commented:
I rearranged those `if` loops in `_math_op_common` in an effort to simplify them a bit.  I don't think I've altered the behaviour.

I'm confused about the in-place case though:  if both `in_place` and `skeleton_cube` are `True`, do we gain anything from creating `new_cube` as a new metadataless cube?  Wouldn't it be more efficient to just set `new_cube = cube`, as in the non-skeleton case?

rcomer commented:
Or...

maybe it would be better to modify the first cube's data within `_binary_op_common`, **before** the call to `_math_op_common`?

```python
if not cube.has_lazy_data():
    if iris._lazy_data.is_lazy_data(rhs):
        cube.data = cube.lazy_data()
    elif in_place and ma.is_masked(rhs) and not isinstance(cube.data, ma.MaskedArray):
        # If lhs is a plain numpy array, we cannot do a straight in-place
        # operation without stripping the mask from rhs.
        cube.data = ma.array(cube.data)
```
Then I wouldn't have introduced a proliferation of `_binary_op_common`-specific logic into `_math_op_common`.

bjlittle commented:
@rcomer No worries, this is all great... perhaps we can pair develop on this together to get it over the line.

What do you say? :smile:

rcomer commented:
Reverted whatsnew as it was now out of date in both content and style and would only lead to git conflicts.

trexfeathers commented:
Since this isn't essential for the v3.0 release candidate, will be revisiting between the release candidate and the full release.

bjlittle commented:
@rcomer I've move this to the `iris` `3.1` milestone, apologies but I don't think we can squeeze this into the `3.0`.

So would you mind targeting `master` instead? Thanks :smile:

rcomer commented:
I've just opened an issue over at https://github.com/dask/dask/issues/6734.

rcomer commented:
It seems the problem is actually with `numpy.ma` https://github.com/numpy/numpy/issues/15200.

rcomer commented:
Rebased 🐱‍👤

rcomer commented:
So https://github.com/numpy/numpy/pull/16022 has seen some movement and, assuming it goes in,  I believe that will fix the issue I was originally trying to work around.  However, we will still need something in Iris for the in-place cases:

These array operations shouldn’t be expected to work
```python
plain_numpy_array *= masked_array
any_numpy_array *= dask_array
```
but I think the equivalent operations on cubes containing these arrays should work.  So I think this PR (or one like it) will still be relevant.

rcomer commented:
If we were only considering in-place cases, I think modifying `cube` before passing it to `_math_op_common` would be the simple/intuitive solution. 🤔

```python
if in_place and not cube.has_lazy_data():
    if iris._lazy_data.is_lazy_data(rhs):
        cube.data = cube.lazy_data()
    elif ma.is_masked(rhs) and not isinstance(cube.data, ma.MaskedArray):
        cube.data = ma.array(cube.data)
```

rcomer commented:
@bjlittle I've had a re-think.  What do you reckon?

* `_math_op_common` now unchanged from `master`.
*  Simple fix for in-place cases.
* Temporary workaround for original issue is self-contained and clearly labelled as temporary, so easy for future devs to remove.

<details>
<summary>Adapting that script we looked at before so it covers all the permutations:</summary>

```python
import itertools

import iris
import dask
import dask.array as da

print('Iris version', iris.__version__)
print('Dask version', dask.__version__)


def multiply_example(first_masked=False, first_lazy=False, second_lazy=False,
                     in_place=False):
    
    if first_masked:
        cube1 = iris.cube.Cube(da.ma.masked_array(range(5), mask=[0, 1, 1, 0, 0]))
    else:
        cube1 = iris.cube.Cube(da.array(range(5)))
        
    cube2 = iris.cube.Cube(da.ma.masked_array(range(5), mask=[0, 0, 1, 0, 1]))
    
    if first_lazy:
        str1 = "First lazy, "
    else:
        str1 = "First real, "
        cube1.data
        
    if first_masked:
        str1 += "masked; "
    else:
        str1 += "not masked; "
        
    if second_lazy:
        str2 = "second lazy."
    else:
        str2 = "second real."
        cube2.data
        
    msg = str1 + str2
    
    if in_place:
        try:
            cube1 *= cube2
        except NotImplementedError as e:
            print(e)
            print('')
            return
        
        result = cube1
        msg += ' Checking in-place'
    else:
        result = cube1 * cube2
        
    print(msg)
    print('Lazy result:', result.has_lazy_data())
    print('Result:', result.data)
    print('')
   
   
for in_place, first_masked, first_lazy, second_lazy in itertools.product([False, True], repeat=4):
    multiply_example(first_masked, first_lazy, second_lazy, in_place)
```
</details>

<details>
<summary>we now get all the right masks in all the right places</summary>

```
Iris version 3.1.dev0
Dask version 2021.04.0

First real, not masked; second real.
Lazy result: False
Result: [0 1 -- 9 --]

First real, not masked; second lazy.
Lazy result: True
Result: [0 1 -- 9 --]

First lazy, not masked; second real.
Lazy result: True
Result: [0 1 -- 9 --]

First lazy, not masked; second lazy.
Lazy result: True
Result: [0 1 -- 9 --]

First real, masked; second real.
Lazy result: False
Result: [0 -- -- 9 --]

First real, masked; second lazy.
Lazy result: True
Result: [0 -- -- 9 --]

First lazy, masked; second real.
Lazy result: True
Result: [0 -- -- 9 --]

First lazy, masked; second lazy.
Lazy result: True
Result: [0 -- -- 9 --]

First real, not masked; second real. Checking in-place
Lazy result: False
Result: [0 1 -- 9 --]

First real, not masked; second lazy. Checking in-place
Lazy result: True
Result: [0 1 -- 9 --]

First lazy, not masked; second real. Checking in-place
Lazy result: True
Result: [0 1 -- 9 --]

First lazy, not masked; second lazy. Checking in-place
Lazy result: True
Result: [0 1 -- 9 --]

First real, masked; second real. Checking in-place
Lazy result: False
Result: [0 -- -- 9 --]

First real, masked; second lazy. Checking in-place
Lazy result: True
Result: [0 -- -- 9 --]

First lazy, masked; second real. Checking in-place
Lazy result: True
Result: [0 -- -- 9 --]

First lazy, masked; second lazy. Checking in-place
Lazy result: True
Result: [0 -- -- 9 --]
```
</summary>

rcomer commented:
Rebased and added a whatsnew.  I think this is about ready to go now.

I'm not 100% happy with the workaround as it seems a bit inefficient to make a new interim cube to pass to `_math_op_common`.  However, this approach does have the advantage of simplicity and hopefully it will only be there temporarily.

There is also still a question of how all this affects `IFunc` and `apply_ufunc`.  I think whether it is a help or hindrance there will depend on exactly which function is being passed, so I'm not clear what could be done about it in general.

pp-mo commented:
Hi @rcomer 
We are now looking to get this into [Iris 3.1.0](https://github.com/SciTools/iris/discussions/4072)
I see you have most recently raised some outstanding questions.
So... is it basically done (good enough) + ready ?   
If so, we can move it into "Awaiting Review" on the [3.1 board](https://github.com/SciTools/iris/projects/26)+ get it signed off !

rcomer commented:
Thanks @pp-mo yes my remaining questions are really ones for the reviewer.  So I have nothing more to add before the review.

rcomer commented:
Rebased and fixed conflicts.

rcomer commented:
https://github.com/numpy/numpy/pull/21977 implemented the NEP13 `__array_ufunc__` dispatch mechanism for masked arrays.  So with numpy main branch we now have:

```python
import numpy
import numpy.ma as ma
import dask
import dask.array as da

print("numpy version:", numpy.__version__)
print("dask version:", dask.__version__)

my_numpy_array = ma.masked_array(range(4), mask=[0, 1, 0, 0])
my_dask_array = da.ma.masked_array(range(4), mask=[0, 0, 1, 0])

result = my_numpy_array * my_dask_array
print(result)
print(result.compute())
```

```
numpy version: 1.24.0.dev0+536.gcb28cd17d
dask version: 2022.05.0
dask.array<multiply, shape=(4,), dtype=int64, chunksize=(4,), chunktype=numpy.MaskedArray>
[0 -- -- 9]
```
:+1: 

So we now only need to handle the in-place cases ourselves :tada:

Have removed the workaround for the non in-place case, and updated the whatsnew and PR title accordingly.

rcomer commented:
These [new xfails](https://github.com/SciTools/iris/runs/7341492773?check_suite_focus=true#step:12:1508) in the test output should disappear when we acquire numpy v1.24 (not yet released).
```
XFAIL lib/iris/tests/unit/analysis/maths/test_add.py::TestMasking::test_partial_mask_second_lazy_not_in_place
  numpy#21977 implemented NEP13 for masked arrays for v1.24
XFAIL lib/iris/tests/unit/analysis/maths/test_divide.py::TestMasking::test_partial_mask_second_lazy_not_in_place
  numpy#21977 implemented NEP13 for masked arrays for v1.24
XFAIL lib/iris/tests/unit/analysis/maths/test_multiply.py::TestMasking::test_partial_mask_second_lazy_not_in_place
  numpy#21977 implemented NEP13 for masked arrays for v1.24
XFAIL lib/iris/tests/unit/analysis/maths/test_subtract.py::TestMasking::test_partial_mask_second_lazy_not_in_place
  numpy#21977 implemented NEP13 for masked arrays for v1.24
```

trexfeathers commented:
@bjlittle do you still want to be assigned to this?

bjlittle commented on lib/iris/tests/unit/analysis/maths/__init__.py:
```diff
@@ -201,12 +202,16 @@ def cube_func(self):
         dat_b = ma.array([2.0, 2.0, 2.0, 2.0], mask=[1, 1, 0, 0])
 
-        cube_a = Cube(dat_a)
+        if second_lazy:
+            cube_a = Cube(da.from_array(dat_a))
+        else:
+            cube_a = Cube(dat_a)
+
         cube_b = Cube(dat_b)
```
```suggestion
        if second_lazy:
            cube_b = Cube(da.from_array(dat_b))
        else:
            cube_b = Cube(dat_b)

        cube_a = Cube(dat_a)
```

rcomer replied:
Have now switched a and b
https://github.com/SciTools/iris/pull/3790/commits/18e2ff8f22989f12dbe0a3b050a37819b64811d6

bjlittle reviewed (approved):
@rcomer Cool beans!

Just as well you weren't holding your breath on this one :wink: 

Lovely job! :100: :partying_face: :beers:

rcomer commented:
Thanks @bjlittle 🎉 🥂
